graphs/lca.cpp

```diff
@@ -1,53 +1,65 @@
+/*
+ * Description: Lowest Common Ancestor (LCA) implemented with Binary Lifting.
+ * Supports functionality like kth Parent, distance between pair of nodes in
+ * number of edges, and distance between pair of nodes in sum of edge weight.
+ * Time: O(n * log(n)) for construction and all queries can be answered in
+ * O(log(n))
+ */
+
+
 struct lca {
+	typedef long long ll;
 	int Log;
 	vector<vector<int>> memo;
-	vector<int> depth;
-
-	lca(const vector<vector<int>> &graph, int root) {//0 - based nodes
-		int n = graph.size();
-		Log = 1;
-		while((1<<Log) < n) ++Log;
-		depth.resize(n,0);
-		memo.resize(n,vector<int>(Log));
-		dfs(root,root,graph);
-	}
+	vector<int> depth, dist;
 
-	void dfs(int node, int par, const vector<vector<int>> &graph) {
+	// use weights of 1 for unweighted tree
+	lca(const vector<vector<pair<int, ll>>> &graph, int root) :
+		depth(graph.size()), dist(graph.size()) {//0 - based nodes
+			int n = graph.size();
+			Log = 1;
+			while((1<<Log) < n) ++Log;
+			memo.resize(n,vector<int>(Log));
+			dfs(root,root,graph);
+		}
+
+	void dfs(int node, int par, const vector<vector<pair<int, ll>>> &graph) {
 		memo[node][0] = par;
-		for(int i = 1; i < Log; ++i) {
+		for(int i = 1; i < Log; ++i)
 			memo[node][i] = memo[memo[node][i-1]][i-1];
-		}
-		for(int to : graph[node]) {
+		for(auto [to, w] : graph[node]) {
 			if(to == par) continue;
 			depth[to] = 1 + depth[node];
+			dist[to] = w + dist[node];
 			dfs(to, node, graph);
 		}
 	}
 
 	int kthPar(int node, int k) {
```

Comment:
make const

Reply:
done - submission below

Reply:
I meant make the function const:
int kthPar(...) const {

Reply:
okay sounds good